### DISCUSSION TITLE
LDAP not yet initialized

### DISCUSSION THREAD

**Discussion by ultimatejedi** [gravitational/teleport#29071] [Q&A] [answered]
I am working on setting up a local instance of Teleport with Active Directory.  When I start the Desktop Service, I get " WARN [WINDOWS_D] skipping desktop discovery: LDAP not yet initialized pid:18134.1 desktop/discovery.go:93"

I found the article on the troubleshooting page (https://goteleport.com/docs/desktop-access/troubleshooting/) with this exact message and following the steps in the solution, my certificate does show up.  I exported it again from the Auth service and reimported it on the DC as the instructions indicate.  I am still getting the same warning when I run "journalctl -fu teleport".  Are there logs elsewhere that could point me in the right direction?

**Comment by ultimatejedi** [comment]
I have also verified that I can connect over LDAPS on port 636 from the DC itself and another windows machine.

**Reply by CiscoNCS** [reply]
Hi,

Can you post logs please also is this community edition or enterprise? Maybe also post a redacted version of the config you are using would be very helpful.

Thanks

**Reply by ultimatejedi** [reply]
Good Morning-

Here is the teleport server config:

```
version: v3
teleport:
  nodename: teleport
  data_dir: /var/lib/teleport
  log:
    output: stderr
    severity: INFO
    format:
      output: text
  ca_pin: ""
  diag_addr: ""
auth_service:
  enabled: "yes"
  listen_addr: 0.0.0.0:3025
  cluster_name: teleport.domain.com
  proxy_listener_mode: multiplex
ssh_service:
  enabled: "yes"
  commands:
  - name: hostname
    command: [hostname]
    period: 1m0s
proxy_service:
  enabled: "yes"
  web_listen_addr: 0.0.0.0:443
  public_addr: teleport.domain.com:443
  https_keypairs: []
  https_keypairs_reload_interval: 0s
  acme:
    enabled: "yes"
    email: email@example.com
```


Here is the node config:

```
version: v3
teleport:
  auth_token: some_token
  proxy_server: teleport.domain.com:443

auth_service:
  enabled: no
ssh_service:
  enabled: no
proxy_service:
  enabled: no

windows_desktop_service:
  enabled: yes
  ldap:
    addr: 'SERVER.domain.local:636'
    domain:   'domain.local'
    username: 'DOMAIN\svc-teleport'
    sid: 'some_sid'
    server_name: 'server.domain.local server.domain.local'
    insecure_skip_verify: false
    ldap_ca_cert: |
        -----BEGIN CERTIFICATE-----
    cert_string
        -----END CERTIFICATE-----

  discovery:
    base_dn: '*'
  labels:
    teleport.internal/resource-id: some_guid
```

Here are the logs from the teleport server:

```
Jul 17 12:54:13 teleport teleport[22164]: ERROR REPORT:
Jul 17 12:54:13 teleport teleport[22164]: Original Error: *errors.errorString acme/autocert: missing server name
Jul 17 12:54:13 teleport teleport[22164]: Stack Trace:
Jul 17 12:54:13 teleport teleport[22164]:         github.com/gravitational/teleport/lib/srv/alpnproxy/proxy.go:392 github.com/gravitational/teleport/lib/srv/alpnproxy.(*Proxy).handleConn
Jul 17 12:54:13 teleport teleport[22164]:         github.com/gravitational/teleport/lib/srv/alpnproxy/proxy.go:326 github.com/gravitational/teleport/lib/srv/alpnproxy.(*Proxy).Serve.func1
Jul 17 12:54:13 teleport teleport[22164]:         runtime/asm_amd64.s:1598 runtime.goexit
Jul 17 12:54:13 teleport teleport[22164]: User Message: acme/autocert: missing server name] alpnproxy/proxy.go:337
Jul 17 13:03:19 teleport teleport[22164]: 2023-07-17T13:03:19Z INFO [CA]        Generating TLS certificate CN=svc-teleport dns_names:[] key_usage:1 not_after:2023-07-17 21:03:19.376743149 +0000 UTC tlsca/ca.go:1111
Jul 17 13:13:19 teleport teleport[22164]: 2023-07-17T13:13:19Z INFO [CA]        Generating TLS certificate CN=svc-teleport dns_names:[] key_usage:1 not_after:2023-07-17 21:13:19.478807771 +0000 UTC tlsca/ca.go:1111
Jul 17 13:18:21 teleport teleport[22164]: 2023-07-17T13:18:21Z WARN [ALPN:PROX] Failed to handle client connection: acme can't get a cert for domain domain.com, add it to the proxy_service.public_addr, or use one of the domains: teleport.domain.com alpnproxy/proxy.go:332
```


Here are the node logs from the teleport node:

```
sudo journalctl -fu teleport
Jul 17 13:13:19 teleportnode teleport[18162]: 2023-07-17T13:13:19Z [WINDOWS_D] ERRO couldn't renew certificate for LDAP auth error:[
Jul 17 13:13:19 teleportnode teleport[18162]: ERROR REPORT:
Jul 17 13:13:19 teleportnode teleport[18162]: Original Error: *ldap.Error LDAP Result Code 200 &#34;Network Error&#34;: tls: failed to verify certificate: x509: certificate is valid for SERVER.domain.local, domain.local, DOMAIN, not SERVER.domain.local SERVER.domain.local
Jul 17 13:13:19 teleportnode teleport[18162]: Stack Trace:
Jul 17 13:13:19 teleportnode teleport[18162]:         github.com/gravitational/teleport/lib/srv/desktop/windows_server.go:528 github.com/gravitational/teleport/lib/srv/desktop.(*WindowsService).initializeLDAP
Jul 17 13:13:19 teleportnode teleport[18162]:         github.com/gravitational/teleport/lib/srv/desktop/windows_server.go:568 github.com/gravitational/teleport/lib/srv/desktop.(*WindowsService).scheduleNextLDAPCertRenewalLocked.func1
Jul 17 13:13:19 teleportnode teleport[18162]:         runtime/asm_amd64.s:1598 runtime.goexit
Jul 17 13:13:19 teleportnode teleport[18162]: User Message: dial
Jul 17 13:13:19 teleportnode teleport[18162]:         LDAP Result Code 200 &#34;Network Error&#34;: tls: failed to verify certificate: x509: certificate is valid for SERVER.domain.local, domain.local, DOMAIN, not SERVER.domain.local SERVER.domain.local] pid:18162.1 desktop/windows_server.go:569
Jul 17 13:16:17 teleportnode teleport[18162]: 2023-07-17T13:16:17Z [WINDOWS_D] WARN skipping desktop discovery: LDAP not yet initialized pid:18162.1 desktop/discovery.go:93
```


As I am posting this, I am noticing on the last line of the server log, the domain does not match the teleport.domain.com.  It is a completely different domain.  I am looking into that.  But I feel like that is a different issue than the node issue and the domain certificate issue.  I have followed the instructions on generating the domain cert.  It does exist and I have verified that it is showing as trusted in the domain.

**Reply by CiscoNCS** [reply]
Do you have a PKI running or is this self signed? also what distro of linux are you running teleport on.

**Reply by CiscoNCS** [reply]
server_name: 'server.domain.local server.domain.local' < this can be removed only useful if you are using IP address instead of FQDN for the addr parameter

**Reply by zmb3** [reply]
> x509: certificate is valid for SERVER.domain.local, domain.local, DOMAIN,

This is the problem. Teleport will not be able to start desktop sessions until it successfully establishes a TLS connection to the LDAP server. Your server is presenting a cert that is valid for a different domain than what you are trying to connect to.

You can fix this by setting the `server_name` property in your config to match the server name presented by the server.

**Reply by CiscoNCS** [reply]
the addr in the config is matching the subject names? i presume he doesn't have the CA added to the trusted chain but that could also work /shrug

**Reply by zmb3** [reply]
1. The values in the config are clearly redacted examples, so we can't say for sure whether they match.
2. It doesn't matter what `addr` is set to in this case, if `server_name` is set, that's what will be used in the TLS handshake.

**Reply by ultimatejedi** [reply]
So I commented out the server_name line and restarted the service on the node machine as it has the same server FQDN listed twice separated by a space (it was generated this was from the implementation script).  This is what I have now:

```
sudo journalctl -fu teleport
Jul 17 16:30:28 teleportnode teleport[23257]: 2023-07-17T16:30:28Z INFO [UPLOAD:1]  Creating directory /var/lib/teleport/log. pid:23257.1 service/service.go:2697
Jul 17 16:30:28 teleportnode teleport[23257]: 2023-07-17T16:30:28Z INFO [UPLOAD:1]  Creating directory /var/lib/teleport/log/upload. pid:23257.1 service/service.go:2697
Jul 17 16:30:28 teleportnode teleport[23257]: 2023-07-17T16:30:28Z INFO [UPLOAD:1]  Creating directory /var/lib/teleport/log/upload/streaming. pid:23257.1 service/service.go:2697
Jul 17 16:30:28 teleportnode teleport[23257]: 2023-07-17T16:30:28Z INFO [UPLOAD:1]  Creating directory /var/lib/teleport/log/upload/streaming/default. pid:23257.1 service/service.go:2697
Jul 17 16:30:28 teleportnode teleport[23257]: 2023-07-17T16:30:28Z INFO [UPLOAD:1]  Creating directory /var/lib/teleport/log. pid:23257.1 service/service.go:2697
Jul 17 16:30:28 teleportnode teleport[23257]: 2023-07-17T16:30:28Z INFO [UPLOAD:1]  Creating directory /var/lib/teleport/log/upload. pid:23257.1 service/service.go:2697
Jul 17 16:30:28 teleportnode teleport[23257]: 2023-07-17T16:30:28Z INFO [UPLOAD:1]  Creating directory /var/lib/teleport/log/upload/corrupted. pid:23257.1 service/service.go:2697
Jul 17 16:30:28 teleportnode teleport[23257]: 2023-07-17T16:30:28Z INFO [UPLOAD:1]  Creating directory /var/lib/teleport/log/upload/corrupted/default. pid:23257.1 service/service.go:2697
Jul 17 16:30:28 teleportnode teleport[23257]: 2023-07-17T16:30:28Z INFO [PROC:1]    The new service has started successfully. Starting syncing rotation status with period 10m0s. pid:23257.1 service/connect.go:811
Jul 17 16:30:28 teleportnode teleport[23257]: 2023-07-17T16:30:28Z INFO [PROC:COMP] upload completer will run every 5m0s events/complete.go:143
Jul 17 16:30:43 teleportnode teleport[23257]: 2023-07-17T16:30:43Z WARN [WINDOWS_D] skipping desktop discovery: LDAP not yet initialized pid:23257.1 desktop/discovery.go:93
```
Both server and node are running Ubuntu Server 22.04
This is a certificate issued by a CA on my Active Directory domain.

I appreciate everyones' help!

**Reply by ultimatejedi** [reply]
I am still receiving "LDAP not yet initialized."

**Reply by CiscoNCS** [reply]
Are you using AD CS by any chance? also is this tier 2/3 setup or?

**Reply by zmb3** [reply]
@ultimatejedi the answer hasn't changed. You are unable to establish a TLS connection. To fix it, you need to set the `server_name` field appropriately.

`server.domain.local server.domain.local` is not a valid hostname.

**Reply by ultimatejedi** [reply]
@zmb3 I commented that line out as recommended by @CiscoNCS above as it was un-needed.  That did resolve the error regarding the certificate as indicated by my most recent long.  However, I am still receiving the warning regarding the LDAP not initialized.

@CiscoNCS: I am using AD CS.  That is how I issued the certificate for my Domain Controller.

If I copy the LDAP_CA_CERT text to a file and name it .cer on my windows machine, it does show as a valid ca certificate and is trusted by my domain.

**Reply by CiscoNCS** [reply]
@ultimatejedi have you added the root and intermediate of ur CA to ubuntu or no?

**Reply by CiscoNCS** [reply]
single tier = just root CA two tier = root & intermediate CA

**Reply by ultimatejedi** [reply]
Now that you mention that, I do not believe I have.  I assumed that was done with the line of code ldap_ca_cert:.

If that is not the case, then how would I go about adding?

**Reply by ultimatejedi** [reply]
I believe it would just be a single tier.

**Reply by CiscoNCS** [reply]
@ultimatejedi convert cer to crt using following "openssl x509 -inform PEM -in <filepath>/certificate.cer -out certificate.crt" you wanna copy that crt file to path of "/usr/local/share/ca-certificates/" after "chmod 775 the crt file" finally run "sudo update-ca-certificates" should add the new CA to the trusted store, open run type pkiview.msc on the windows issuing server it'll show you the hierarchy.

**Reply by ultimatejedi** [reply]
OK.  I did that, shows that it added the certificate successfully.  Restarted the service and still getting the LDAP not yet initialized.

I did check the pkiview and only a single tier.

**Reply by CiscoNCS** [reply]
when you added the ldaps cert to the domain controller did you import it like this MMC >> Certificates >> Services >> AD DS >> Personal Store ?

**Reply by CiscoNCS** [reply]
secondly in the subject name its DNS/FQDN not just "DC01"

**Reply by ultimatejedi** [reply]
To make sure I am clear, I ran this on the teleport node that I am trying to setup with the Windows Desktop Server.  Not on the main teleport server.

**Reply by CiscoNCS** [reply]
dont know if i'm breaking rules here but you got discord? we can take this convo off github for better troubleshooting if you'd like because i got a full teleport node working with windows desktop ssh DB and kube i'm fairly interested whats different about urs

**Reply by CiscoNCS** [reply]
added! vlan

**Reply by ultimatejedi** [reply, marked as answer]
Update.  I was able to resolve the issue when I discovered that the svr-teleport AD account that was created by the install script was disabled.  When I went to enable it, I received a warning that the auto-generated password did not meet the complexity requirements.  I changed the password, and enabled.  AD connection now works as expected.  Thanks everyone for your assistance!

**Reply by savv3** [reply]
Thanks for the comment. This solved my issue.
